File: code/java/ch05/src/main/java/com/github/zxh0/luago/state/LuaValue.java
package com.github.zxh0.luago.state;

import com.github.zxh0.luago.api.LuaType;
import com.github.zxh0.luago.number.LuaNumber;

import static com.github.zxh0.luago.api.LuaType.*;

class LuaValue {

    static LuaType typeOf(Object val) {
        if (val == null) {
            return LUA_TNIL;
        } else if (val instanceof Boolean) {
            return LUA_TBOOLEAN;
        } else if (val instanceof Long || val instanceof Double) {
            return LUA_TNUMBER;
        } else if (val instanceof String) {
            return LUA_TSTRING;
        } else {
            throw new RuntimeException("TODO");
        }
    }

    static boolean toBoolean(Object val) {
        if (val == null) {
            return false;
        } else if (val instanceof Boolean) {
            return (Boolean) val;
        } else {
            return true;
        }
    }

    // http://www.lua.org/manual/5.3/manual.html#3.4.3
    static Double toFloat(Object val) {
        if (val instanceof Double) {
            return (Double) val;
        } else if (val instanceof Long) {
            return ((Long) val).doubleValue();
        } else if (val instanceof String) {
            return LuaNumber.parseFloat((String) val);
        } else {
            return null;
        }
    }

    // http://www.lua.org/manual/5.3/manual.html#3.4.3
    static Long toInteger(Object val) {
        if (val instanceof Long) {
            return (Long) val;
        } else if (val instanceof Double) {
            double n = (Double) val;
            return LuaNumber.isInteger(n) ? (long) n : null;
        } else if (val instanceof String) {
            return toInteger((String) val);
        } else {
            return null;
        }
    }

    private static Long toInteger(String s) {
        Long i = LuaNumber.parseInteger(s);
        if (i != null) {
            return i;
        }
        Double f = LuaNumber.parseFloat(s);
        if (f != null && LuaNumber.isInteger(f)) {
            return f.longValue();
        }
        return null;
    }

}

File: code/csharp/ch07/luavm/state/LuaValue.cs
using System;
using System.Collections.Generic;
using luavm.api;
using LuaType = System.Int32;

namespace luavm.state
{
    public class LuaValue
    {
        internal readonly object value;

        public LuaValue(object value)
        {
            this.value = value;
        }

        public LuaTable toLuaTable()
        {
            return (LuaTable) value;
        }

        public string toString()
        {
            return Convert.ToString(value);
        }

        public long toInteger()
        {
            return Convert.ToInt64(value);
        }

        public double toFloat()
        {
            return Convert.ToDouble(value);
        }

        public bool isString()
        {
            return value.GetType().IsEquivalentTo(typeof(string));
        }

        public bool isLuaTable()
        {
            return value.GetType().IsEquivalentTo(typeof(LuaTable));
        }

        public bool isFloat()
        {
            return value.GetType().IsEquivalentTo(typeof(double));
        }

        public bool isInteger()
        {
            return value.GetType().IsEquivalentTo(typeof(long));
        }

        internal LuaType typeOf()
        {
            if (value == null)
            {
                return Consts.LUA_TNIL;
            }

            // Console.WriteLine("\n" + value.GetType().Name);

            switch (value.GetType().Name)
            {
                case "Boolean": return Consts.LUA_TBOOLEAN;
                case "Double": return Consts.LUA_TNUMBER;
                case "Int64": return Consts.LUA_TNUMBER;
                case "String": return Consts.LUA_TSTRING;
                case "LuaTable": return Consts.LUA_TTABLE;
               }

            throw new Exception("todo!");
        }

        internal static Tuple<double, bool> convertToFloat(LuaValue luaValue)
        {
            switch (luaValue.value.GetType().Name)
            {
                case "Double": return Tuple.Create(luaValue.toFloat(), true);
                case "Int64": return Tuple.Create(luaValue.toFloat(), true);
                case "String": return number.Parser.ParseFloat(Convert.ToString(luaValue.value));
                default: return Tuple.Create(0d, false);
            }
        }

        internal static Tuple<long, bool> convertToInteger(object val)
        {
            switch (val.GetType().Name)
            {
                case "Int64": return Tuple.Create((long) val, true);
                case "Double": return number.Math.FloatToInteger((double) val);
                case "String": return Tuple.Create(Convert.ToInt64(val), true);
                default: return Tuple.Create(0L, false);
            }
        }

        private Tuple<long, bool> _stringToInteger(string s)
        {
            var v = number.Parser.ParseInteger(s);
            if (v.Item2)
            {
                return Tuple.Create(v.Item1, true);
            }

            var v2 = number.Parser.ParseFloat(s);
            if (v2.Item2)
            {
                return number.Math.FloatToInteger(v2.Item1);
            }

            return Tuple.Create(0L, false);
        }
    }
}